include/ua_plugin_history_data_service.h

```diff
@@ -0,0 +1,79 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ *    Copyright 2018 (c) basysKom GmbH <opensource@basyskom.com> (Author: Peter Rustler)
+ */
+
+#ifndef UA_PLUGIN_HISTORY_DATA_SERVICE_H_
+#define UA_PLUGIN_HISTORY_DATA_SERVICE_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "ua_types.h"
+#include "ua_server.h"
+
+struct UA_HistoryDataService;
+typedef struct UA_HistoryDataService UA_HistoryDataService;
+
+struct UA_HistoryDataService {
+    void *context;
+
+    void
+    (*deleteMembers)(UA_HistoryDataService *service);
+
+    /* This function will be called when a node value is set.
+     * Use this to insert data into your database(s) if polling is not suitable
+     * and you need to get all data changes.
+     * Set it to NULL if you do not need it.
+     *
+     * context is the context of the UA_HistoryDataService. UA_HistoryDataService.context
+     * server is the server this node lives in.
+     * sessionId and sessionContext identify the session which set this value.
+     * nodeId the node id for which data was set.
+     * value the new value.
+     */
+    void
+    (*setValue)(void *context,
+                UA_Server *server,
+                const UA_NodeId *sessionId,
+                void *sessionContext,
+                const UA_NodeId *nodeId,
+                const UA_DataValue *value);
+
+    /* This function is called if a history read is requested
+     * with isRawReadModified set to false.
+     * Setting it to NULL will result in a response with statuscode UA_STATUSCODE_BADHISTORYOPERATIONUNSUPPORTED.
+     *
+     * context is the context of the UA_HistoryDataService. UA_HistoryDataService.context
+     * server is the server this node lives in.
+     * sessionId and sessionContext identify the session which set this value.
+     * nodeId the node id for which data was set.
+     * request is the complete request from the client.
+     * details is the details object which is also present in the request as an extension object.
+     * response the response to fill for the client.
+     * results is a pointer with an already allocated array of UA_HistoryReadResult with
+     *         size of request->nodesToReadSize.
+     *         This is also present in the response.
+     */
+    void
+    (*readRaw)(void *context,
+               UA_Server *server,
+               const UA_NodeId *sessionId,
+               void *sessionContext,
+               const UA_HistoryReadRequest *request,
+               const UA_ReadRawModifiedDetails *details,
+               UA_HistoryReadResponse *response,
+               UA_HistoryReadResult* results);
```

Comment:
The `results` parameter is a member of `response` (as you correctly state) and therefore could be omitted as a additional parameter. It might confuse while implementing the plugin function.

Reply:
Yes, the results object is already in the response. But to make t easier for the programmer it is here, also. In the response it is an extrension object and do not have the type. To show the programmer that we expect a UA_HistoryReadResult array to be filled, it is here.

Reply:
Fair enough. Totally forgot, that the response actually is an extension object. :+1:

Reply:
Hmm. I see the point.

On the other hand, having the same "pointer" twice in the arguments can be very confusing for users.

Reply:
It is not the same pointer twice.
You have one untyped (void*) pointer as a member in the response.
And you have the results pointer which has the proper type and allocated memory you just have to fill.

That makes it much harder for a programmer to return the wrong type in the response.